compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEHoistLogicalObjFifo.cpp

```diff
@@ -0,0 +1,60 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+
+#define DEBUG_TYPE "iree-amdaie-hoist-logical-objectfifo"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// Hoist logical objectFifo operations until one of the operands is located
+/// within the same scope.
+LogicalResult hoistLogicalObjFifoOp(RewriterBase &rewriter,
+                                    AMDAIE::LogicalObjectFifoFromMemrefOp op) {
+  Operation *parentOp = op;
+  while (parentOp) {
+    Operation *newParentOp = parentOp->getParentOp();
+    if (llvm::any_of(op->getOperands(), [&](Value operand) {
+          return operand.getDefiningOp() &&
+                 newParentOp->isProperAncestor(operand.getDefiningOp());
+        })) {
+      break;
+    }
+    if (isa<AMDAIE::WorkgroupOp, AMDAIE::ControlCodeOp, func::FuncOp>(
+            newParentOp)) {
+      break;
+    }
+    parentOp = newParentOp;
+  }
+  if (parentOp && parentOp != op) rewriter.moveOpBefore(op, parentOp);
+  return failure();
+}
+
+namespace {
+struct AMDAIEHoistLogicalObjFifoPass
+    : public impl::AMDAIEHoistLogicalObjFifoBase<
+          AMDAIEHoistLogicalObjFifoPass> {
+  void runOnOperation() override;
+};
+
+void AMDAIEHoistLogicalObjFifoPass::runOnOperation() {
+  Operation *parentOp = getOperation();
+  IRRewriter rewriter(parentOp->getContext());
+  parentOp->walk([&](AMDAIE::LogicalObjectFifoFromMemrefOp op) {
+    (void)hoistLogicalObjFifoOp(rewriter, op);
```

Comment:
I'm not sure if `walk` enumerates all the ops it will visit before hand, I don't think so, so I think you actually visit all the hoisted ops twice (once after hoisting) here. 

FWIW I find code that
1) walk to find targets
2) process targets

easier to grok.

Reply:
> so I think you actually visit all the hoisted ops twice (once after hoisting) 

The hoisted ops will be inserted before the ancestor op, so should not be visited again?

Reply:
Oh yes, you're right.